EXR: support mipmaps in exr image files

aras-p opened the pull request:
Tiled EXR files can have mipmaps levels. When they do, and the odd size rounding mode matches what all the graphics APIs do ("round down"), load them when loading the exr file. Using the same machinery that the DDS mipmap loading uses.

`Bonita.exr` from OpenEXR test images, for example, now loads with the mip levels: https://github.com/AcademySoftwareFoundation/openexr-images/blob/main/MultiResolution/Bonita.exr

I've also tested on Animal Lab "ALab" USD data set which has exrs with mipmaps. Seems to work!

aras-p committed:
EXR: support mipmaps in exr image files

Tiled EXR files can have mipmaps levels. When they do,
and the odd size rounding mode matches what all the
graphics APIs do ("round down"), load them when loading
the exr file. Using the same machinery that the DDS
mipmap loading uses.

baldurk commented on renderdoc/core/image_viewer.cpp:
```diff
@@ -748,15 +749,28 @@ void ImageViewer::RefreshFile()
+      }
     }
+    read_data.mips = mipCount;
+    texDetails.mips = mipCount;
```
Can you set the other members of `read_data` here? At the moment they shouldn't be used but I'd rather have them initialised properly since they're simple to get at, in case they're used in future as this path is very unlikely to be tested often.

baldurk commented on renderdoc/core/image_viewer.cpp:
```diff
@@ -897,9 +915,7 @@ void ImageViewer::RefreshFile()
-  m_FrameRecord.frameInfo.uncompressedFileSize = datasize;
-
-  read_dds_data read_data = {};
+  m_FrameRecord.frameInfo.uncompressedFileSize = std::max(datasize, read_data.buffer.size());
```
`RDCMAX` rather than `std::max` please, same thing basically but fewer stl dependencies :).

baldurk commented on renderdoc/core/image_viewer.cpp:
```diff
@@ -637,6 +637,7 @@ void ImageViewer::RefreshFile()
   size_t datasize = 0;
 
   bool dds = false;
+  read_dds_data read_data = {};
```
I'm going to be annoying and say that I think this is the right approach but it should be refactored so we're not quite so silently hijacking the dds path. This code has been DDS specific because historically DDS files are the only ones with mips, slices, or complex formats, so if we're adding a new format we should untangle that more clearly.

What I'd suggest is:

* Rename `dds` -> `tex` in the structures, add a comment that this is for 'texture' type images - ie. any file format that is more complex in some dimension than one rectangle of pixels in a regular Nbpp float or normalised format.
* `is_dds_file`/`load_dds_from_file`/`write_dds_to_file` can stay named the same.
* The header is a bit annoying, there's not a great 'common' place to put the structs. I think I'd slightly prefer adding a new tiny header in the same place for the structs, but I'd also be fine just renaming the header file similarly `dds` -> `tex` and accept that the dds functions will live there for now.
* Refactor the control flow here such that the 'dds' path isn't just for dds files.

aras-p replied:
Yeah I also had a "ick, this is hijacking structs that were dds only before", but I wondered whether a larger "refactor" would be good or not. Will do!

baldurk replied:
It's definitely borderline. It doesn't need a huge refactor I think mostly I'm picturing a rename pass to be clearer so people reading the code have an easier time and don't make bad assumptions, but otherwise the end result probably won't look that different to what you have here.

aras-p replied:
Ok, I think I did your suggestion, and factored out all the EXR loading into a separate function, since it was getting fairly large being just inside RefreshFile. That ok?

baldurk replied:
Certainly a reasonable step, and looks good thanks!

aras-p committed:
Refactor: rename dds_data -> tex_data

And move it out into a tex_data.h header. This is in preparation
for getting mipmaps out of EXR files, where read_dds_data for
that would be a bit confusing.

aras-p committed:
Refactor: is_exr_file checks memory buffer

All places that used is_exr_file already had first four bytes
of the file read into memory. So make is_exr_file work just
like is_dds_file does and check those bytes directly, no need
to hit stdio again.

aras-p committed:
Refactor: factor out load_exr_from_file

Move all of EXR reading/preparation code into load_exr_from_file,
similar to how load_dds_from_file is done.

ImageViewer::RefreshFile code now uses very similar flow
for both DDS and EXR files (i.e. "what can have mipmaps")
